### ISSUE TITLE
response validation slow

### ISSUE THREAD

**Issue by kraf** [apigee-127/swagger-tools#280]
I have several endpoints sending payloads upwards of 50 MB JSON and the response schema validation is taking ages (~ 2 min) even if I only say the response is an object using

```yaml
schema:
  type: object
  properties: {}
```

I see you are using `z-schema` to do the validations. On their own Github page they published a [benchmark](https://rawgit.com/zaggino/z-schema/master/benchmark/results.html) showing that libraries like `ajv` can be more than a 100 times faster. Are you open to change the validation library in use?

I did not find a way to disable validation for certain endpoints and am currently using `res.write` to bypass the checks, see also https://github.com/swagger-api/swagger-node/issues/263. Is there an easier way that I'm overlooking?

**Comment by whitlockjc** [comment]
I'm not against it but have you checked to see if there is indeed a performance increase?  It would be really simple to write an ajv validator and a z-schema validator for your 50MB JSON and report the findings.

**Comment by kraf** [comment]
Thanks for the quick response. I tested it now and you were right to ask, the validation time is negligable, it's just a few ms. I was pretty sure the problem was there but now I'm lost. The response time using `res.write` is ~20s while using `res.json` or `res.send` puts it over 2min.
Do you have any idea what could slow the middleware down with bigger JSON objects?

**Comment by whitlockjc** [comment]
If you turn off all response validation and the use `res.json` or `res.send`, how does it perform?  I think we both know but just to be sure.

**Comment by kraf** [comment]
I rewrote it and commented out the validator middleware in `connect_middleware.js` and it performs well, I get an answer after 18s. If I put in the middleware I'm at 2min again, so the problem is really with the validator. I will try to find some time to CPU profile this which should clear things up.

**Comment by whitlockjc** [comment]
I have nothing close to a 50Mb JSON file but if I can produce one or find one, I'll profile it.

**Comment by whitlockjc** [comment]
I am unable to reproduce the severity of the issue you speak of.  I am testing with a 134Mb file using Connect and swagger-tools and it only took 13s *(`time` shows 19s but the request/response debugging shows 13s)*:

```
time curl -L -o issue-280.json http://localhost:3000/api/issue-280
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  134M    0  134M    0     0  9831k      0 --:--:--  0:00:14 --:--:-- 35.5M
curl -L -o issue-280.json http://localhost:3000/api/issue-280  0.02s user 0.19s system 1% cpu 14.048 total
```

I realize 13s is still long but for 134Mb file it doesn't seem too out of the realm of possibility.  Also, standalone z-schema takes around 8s to validate the same file.  I will now start profiling to see if I can find out where this could be happening.

**Comment by whitlockjc** [comment]
Validate value is only taking ~12% of the response time:

```
68.54 %	20473.2 ms68.54 %	(program):1(idle)	
4933.0 ms16.51 %	8741.3 ms29.26 %	Application.js:4module.exports.issue280	
2374.1 ms7.95 %	3768.1 ms12.61 %	swagger-validator.js:110validateValue	
1109.7 ms3.72 %	1246.3 ms4.17 %	net.js:638Socket._writeGeneric	
650.8 ms2.18 %	650.8 ms2.18 %	(program):1(garbage collector)	
144.6 ms0.48 %	144.6 ms0.48 %	(program):1byteLength	
136.6 ms0.46 %	136.6 ms0.46 %	(program):1writev	
```

`Application#issue280` just returns the JSON file as a string which all servers I'm aware of will do so I'm assuming the heaviest part of this is the large file to JSON processing.

**Comment by kraf** [comment]
I profiled it myself now and figured out that my problem was here: https://github.com/apigee-127/swagger-tools/blob/master/middleware/swagger-validator.js#L125 - this line took >1 min.
I saved the contents of `val` into a file, loaded the file in a REPL and `JSON.parse` only took 8 seconds there. I have no idea what is going on here. Why should there be a difference?

I'm currently forced to use Node v0.10.34 so it being pretty old I tried starting my server with the newest version (v4.1.0) and everything worked fine, the issue was gone.

Thanks for looking into this, you were very helpful! I'm sorry it turned out to be a wild goose chase. I saw that bypassing the validation was giving me good performance and I made a wrong assumption. I'll keep using my workaround until I can update.
